dm/master/server.go

```diff
@@ -1985,3 +2263,341 @@ func (s *Server) workerArgsExtractor(args ...interface{}) (workerrpc.Client, str
+		request := &pb.PurgeWorkerRelayRequest{}
+		if err = request.Unmarshal(operate.Request); err != nil {
+			operate.Err = terror.ErrMasterUnmarshalOperateRequest.Delegate(err, pb.OperateType_PurgeWorkerRelay).Error()
+			break
+		}
+		response, err1 := s.purgeWorkerRelay(ctx, request)
+		if err1 != nil {
+			operate.Err = err1.Error()
+			break
+		}
+		responseBytes, err = response.Marshal()
+	case pb.OperateType_CheckTask:
+		request := &pb.CheckTaskRequest{}
+		if err = request.Unmarshal(operate.Request); err != nil {
+			operate.Err = terror.ErrMasterUnmarshalOperateRequest.Delegate(err, pb.OperateType_CheckTask).Error()
+			break
+		}
+		response, err1 := s.checkTask(ctx, request)
+		if err1 != nil {
+			operate.Err = err1.Error()
+			break
+		}
+		responseBytes, err = response.Marshal()
+	default:
+		log.L().Error("unhandle operate", zap.Reflect("type", operate.Tp))
+		return
+	}
+
+	if err != nil {
+		operate.Err = terror.ErrMasterMarshalResponse.Delegate(err).Error()
+	} else {
+		operate.Response = responseBytes
+	}
+
+	operateBytes, err := operate.Marshal()
+	if err != nil {
+		log.L().Error("marshal operate failed", zap.Error(err))
+		return
+	}
+
+	err = s.etcdClient.Update(ctx, path, string(operateBytes), 0)
+	if err != nil {
+		log.L().Error("update operate failed", zap.Error(err))
+	}
+}
+
+// saveRequestAndWaitResponse saves request to etcd, and wait for the response
+func (s *Server) saveRequestAndWaitResponse(ctx context.Context, tp pb.OperateType, request proto.Message, response proto.Message) error {
+	requestBytes, err := proto.Marshal(request)
+	if err != nil {
+		return terror.ErrMasterMarshalRequest.Delegate(err, request)
+	}
+
+	operateID := getOperateID()
+	operateIDStr := strconv.FormatInt(operateID, 10)
+
+	operate := pb.Operate{
+		Tp:      tp,
+		Request: requestBytes,
+	}
+	opBytes, err := operate.Marshal()
+	if err != nil {
+		return terror.ErrMasterMarshalOperate.Delegate(err, operate)
+	}
+	revision, err := s.etcdClient.Create(ctx, operateIDStr, string(opBytes), nil)
+	if err != nil {
+		return terror.ErrMasterWriteEtcd.Delegate(err)
+	}
+
+	watchPath := path.Join(defaultOperatePath, operateIDStr)
+	watchCh := s.etcdClient.Watch(ctx, watchPath, revision+1)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case wresp := <-watchCh:
+			if wresp.Err() != nil {
+				return terror.ErrMasterWatchEtcd.Delegate(wresp.Err(), watchPath)
+			}
+
+			// should only have one event
+			if len(wresp.Events) > 1 {
+				log.L().Warn("have more than one event on key in etcd", zap.String("key", operateIDStr))
+			}
+
+			for _, ev := range wresp.Events {
+				operate := &pb.Operate{}
+				err := operate.Unmarshal(ev.Kv.Value)
+				if err != nil {
+					return terror.ErrMasterUnmarshalOperate.Delegate(err)
+				}
+
+				if len(operate.Err) != 0 {
+					return errors.New(operate.Err)
+				}
+
+				return terror.ErrMasterUnmarshalResponse.Delegate(proto.Unmarshal(operate.Response, response))
```

Comment:
should we need to delete the key-value after handling the request?

Reply:
I think we can save it for debug now